executors.md

```diff
@@ -0,0 +1,179 @@
+traits and built-in executors. In addition, it sets up a core expectation for all
+tasks that they are able to spawn additional tasks, while giving fine-grained
+control over what executor that spawning is routed to.
+
+NOTE: this RFC assumes that [RFC #2] is accepted.
+
+[RFC #2]: https://github.com/rust-lang-nursery/futures-rfcs/pull/2
+
+# Motivation
+[motivation]: #motivation
+
+This is a follow-up to the [Tokio Reform], [Futures 0.2], and [Task Context]
+RFCs, harmonizing their designs.
+
+[Tokio Reform]: https://github.com/tokio-rs/tokio-rfcs/pull/3
+[Futures 0.2]: https://github.com/rust-lang-nursery/futures-rfcs/pull/1
+[Task Context]: https://github.com/rust-lang-nursery/futures-rfcs/pull/2
+
+The design in this RFC has the following goals:
+
+- Add a core assumption that tasks are *always* able to spawn additional tasks,
+  avoiding the need to reflect this at the API level. (Note: this assumption is
+  only made in contexts that also assume an allocator.)
+
+- Provide fine-grained control over which executor is used to fulfill that assumption.
+
+- Provide reasonable built-in executors for both single-threaded and
+  multithreaded execution.
+
+It brings the futures library into line with the design principles of the [Tokio
+Reform], but also makes some improvements to the `current_thread` design.
+
+# Proposed design
+
+## The core executor abstraction: `Spawn`
+
+First, in `futures-core` we provide a basic definition of a general-purpose
+executor:
+
+```rust
+trait Spawn {
+    fn spawn(&self, f: Box<Future<Item = (), Error = ()> + Send>);
+}
+```
+
+The `Spawn` trait is quite straightforward (some alternatives and tradeoffs are
+discussed in the next section), and crucially is object-safe.
+
+We then build in an executor to the task context, stored internally as a trait
+object, which allows us to provide the following methods:
+
+```rust
+impl task::Context {
+    // Spawn onto the current default executor
+    fn spawn<F>(&self, F)
+        where F: Future<Item = (), Error = ()> + Send + 'static;
+
+    // Run the given closure with a context using a new default executor
+    fn with_spawn<S, F, R>(&mut self, spawn: S, f: F) -> R
+        where S: Spawn, F: FnOnce(&mut task::Context) -> R;
+}
+```
+
+With those APIs in place, we've achieved two of our goals: it's possible for any
+future to spawn a new task, and to exert fine-grained control over generic task
+spawning within a sub-future.
+
+## Built-in executors
+
+In the `futures-executor` crate, we then add two executors: `ThreadPool` and
+`LocalPool`, providing multi-threaded and single-threaded execution,
+respectively.
+
+### Multi-threaded execution: `ThreadPool`
+
+The `ThreadPool` executor works basically like `CpuPool` today:
+
+```rust
+struct ThreadPool { ... }
+impl Spawn for ThreadPool { ... }
+
+impl ThreadPool {
+    // sets up a pool with the default number of threads
+    fn new() -> ThreadPool;
+
+    // starts up the worker threads
+    fn run();
```

Comment:
Is the intended way of using `ThreadPool`, the following:

```rust
let f = ...; // some arbitrarily complex future.
let tp = ThreadPool::new();
tp.run();
let lp = LocalPool::new();
lp.run_until(f, tp);
```

Currently, there's no way to run stuff on the thread pool without blocking the current thread. And `run` seems to make the situation worse, since forgetting to call it here will result in `run_until` never returning as long as something tries to execute another task using spawn.

Reply:
I don't know that `run` is necessary-- it seems like you could have `ThreadPool` configured to start up when the first task is spawned.

Reply:
WRT running on a `LocalPool`, I think that unless you need to spawn `!Sync` tasks, you shouldn't need to create a `LocalPool` to wait on the result of a `Future`. I had imagined modifying `wait` to take a `Spawn` so that you could do something like this:

```rust
let future  ...; // some arbitrarily complex future.
let thread_pool = ThreadPool::new();
let res = future.wait(&thread_pool);
```

This would run `future` on the current thread, but any `spawn`ed tasks would go onto the `thread_pool`.

Reply:
What does that code look like without using `async`?

Reply:
And about `Future::wait`, that's just a convenience function that does what I suggested. Hence it doesn't really answer the underlying question, right?

Reply:
@ahmedcharles 
> What does that code look like without using async?

```rust
let local_pool = LocalPool::new();
let future = create_some_future_that_holds_a_reference_to_local_pool_and_uses_it_to_spawn{&local_pool);

let thread_pool = ThreadPool::new();
let res = local_pool.run_until(future, &thread_pool);
```

> And about Future::wait, that's just a convenience function that does what I suggested. Hence it doesn't really answer the underlying question, right?

Your original question was:

> Currently, there's no way to run stuff on the thread pool without blocking the current thread. And run seems to make the situation worse, since forgetting to call it here will result in run_until never returning as long as something tries to execute another task using spawn.

How do you want to receive the result of your data onto the main thread without blocking it? Do you want to poll for the data? Do you want to receive some kind of thread wakeup? All of these can be accomplished via normal thread-communication mechanisms (channels, mutexes, condition variables, etc).

Reply:
At the end of the day, all programs have to block (for user input) or exit, so in a sense, you're right. There are nuances however.

The C++ Networking TS, for example has 4 groups of functions that progress the executor:

run(timeout) - blocking, runs until no work remains or the timeout expires. (this seems to be what we currently have, without the timeout)
run_one(timeout) - blocking, runs until at most one handler has completed or the timeout expires.
poll - non-blocking, runs all ready handlers and then returns.
poll_one - non-blocking, runs at most one ready handler and returns.

This is enough flexibility that it's possible to mix these options along with a GUI that isn't future aware without much inefficiency. The alternative, which works either way, is to use a future which understands how to bridge the gap between the non-future aware GUI and the rest of the system. With poll/poll_one, the interaction could be as simple as calling poll when the GUI is idle every hundred milliseconds or so. With run(timeout) you could block the GUI thread for at most a hundred milliseconds or so before starting to respond to input again. With run_one (no timeout), you could have a future which returns when a GUI input is received.

With only run, it's all futures, all the time, for everything, which doesn't seem very incremental and hence would be harder to adopt for sufficiently complex applications. Or is there something I'm missing?